jwks/key.go

```diff
@@ -0,0 +1,145 @@
+package jwks
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+)
+
+const typeRSA = "RSA"
+
+var supportedKeyTypes = map[string]struct{}{
+	typeRSA: {},
+}
+
+// JWK defines JSON Web Key structure.
+type JWK struct {
+	Alg string `json:"alg"`           // algorithm
+	Crv string `json:"crv,omitempty"` // curve - for EC keys
+	D   string `json:"d"`             // private exponent
+	DP  string `json:"dp"`            // d mod (p-1)
+	DQ  string `json:"dq"`            // d mod (p-1)
+	E   string `json:"e"`             // public exponent
+	K   string `json:"k,omitempty"`   // symmetric key
+	Kid string `json:"kid"`           // Key ID
+	Kty string `json:"kty"`           // Key Type
+	N   string `json:"n"`             // modulus
+	P   string `json:"p"`             // prime factor 1
+	Q   string `json:"q"`             // prime factor 2
+	QI  string `json:"qi"`            // q^(-1) mod p
+	Use string `json:"use"`
+	X   string `json:"x,omitempty"` // x coordinate - for EC keys
+	Y   string `json:"y,omitempty"` // y coordinate - for EC keys
+}
+
+// DecodePublicKey decodes JWK to public key.
+func (j *JWK) DecodePublicKey() (crypto.PublicKey, error) {
+	if _, ok := supportedKeyTypes[j.Kty]; !ok {
+		return nil, fmt.Errorf("unsupported key type %s", j.Kty)
+	}
+
+	var result interface{}
+
+	if j.Kty == typeRSA {
+		if j.N == "" || j.E == "" {
+			return nil, errors.New("malformed JWK RSA j")
+		}
+
+		e, err := DecodeBase64URLToBigInt(j.E)
+		if err != nil {
+			return nil, errors.New("malformed JWK RSA j")
+		}
+		eBytes := e.Bytes()
+		if len(eBytes) < 4 {
+			ndata := make([]byte, 4)
+			copy(ndata[4-len(eBytes):], eBytes)
+			eBytes = ndata
+		}
+
+		pubKey := &rsa.PublicKey{
+			N: &big.Int{},
+			E: int(binary.BigEndian.Uint32(eBytes)),
+		}
+
+		n, err := DecodeBase64URLToBigInt(j.N)
+		if err != nil {
+			return nil, errors.New("malformed JWK RSA j")
+		}
+		pubKey.N = n
+
+		result = pubKey
+	}
+
+	return result, nil
+}
+
+// DecodePrivateKey decodes JWK to private key.
+func (j *JWK) DecodePrivateKey() (crypto.PrivateKey, error) {
+	if _, ok := supportedKeyTypes[j.Kty]; !ok {
+		return nil, fmt.Errorf("unsupported key type %s", j.Kty)
+	}
+
+	var result interface{}
+	var err error
+
+	if j.Kty == typeRSA {
+		// Decode base64url-encoded JWK components
+		components := []string{j.N, j.E, j.D, j.P, j.Q, j.DP, j.DQ, j.QI}
+		decodedComponents := make([]*big.Int, len(components))
+
+		for i, component := range components {
+			decodedComponents[i], err = DecodeBase64URLToBigInt(component)
```

Comment:
The previous implementation ([gojwk](https://github.com/mendsley/gojwk)) returned an error if the D component was empty ([reference](https://github.com/mendsley/gojwk/blob/4d5ec6e58103388d6cb0d7d72bc72649be4f0504/key.go#L237)). Could you please explain the reasoning behind changing this behavior?

Reply:
Totally right!